src/hooks/init/check-update.ts

```diff
@@ -1,69 +1,135 @@
 import {Hook} from '@oclif/core'
-import type {Chalk} from 'chalk'
-import {spawn} from 'child_process'
-import {readFile, stat} from 'fs/promises'
-import {join} from 'path'
+import chalk from 'chalk'
+import makeDebug from 'debug'
+import {spawn} from 'node:child_process'
+import {Stats} from 'node:fs'
+import {readFile, stat, utimes} from 'node:fs/promises'
+import {dirname, join, resolve} from 'node:path'
+import {fileURLToPath} from 'node:url'
+
+const readJSON = async <T>(file: string): Promise<{contents: T; stat: Stats}> => ({
+  contents: JSON.parse(await readFile(file, 'utf8')),
+  stat: await stat(file),
+})
+
+function hasNotBeenMsSinceDate(ms: number, date: Date): boolean {
+  const now = new Date()
+  const diff = now.getTime() - date.getTime()
+  return diff < ms
+}
+
+function convertToMs(frequency: number, unit: 'days' | 'hours' | 'milliseconds' | 'minutes' | 'seconds'): number {
+  switch (unit) {
+    case 'days': {
+      return frequency * 24 * 60 * 60 * 1000
+    }
+
+    case 'hours': {
+      return frequency * 60 * 60 * 1000
+    }
+
+    case 'milliseconds': {
+      return frequency
+    }
+
+    case 'minutes': {
+      return frequency * 60 * 1000
+    }
+
+    case 'seconds': {
+      return frequency * 1000
+    }
+
+    default: {
+      // default to minutes
+      return frequency * 60 * 1000
+    }
+  }
+}
 
 const hook: Hook<'init'> = async function ({config}) {
+  const debug = makeDebug('update-check')
   const file = join(config.cacheDir, 'version')
 
   // Destructure package.json configuration with defaults
   const {
-    timeoutInDays = 60,
+    authorization = '',
+    frequency,
+    frequencyUnit,
     message = '<%= config.name %> update available from <%= chalk.greenBright(config.version) %> to <%= chalk.greenBright(latest) %>.',
     registry = 'https://registry.npmjs.org',
-    authorization = '',
-  } = (config.pjson.oclif as any)['warn-if-update-available'] || {}
+    timeoutInDays = 60,
+  } = config.pjson.oclif['warn-if-update-available'] || {}
+
+  const warningFrequency = (config.scopedEnvVar('NEW_VERSION_CHECK_FREQ') as number | undefined) ?? frequency
+  const warningFrequencyUnit =
+    (config.scopedEnvVar('NEW_VERSION_CHECK_FREQ_UNIT') as typeof frequencyUnit) ?? frequencyUnit ?? 'minutes'
+
+  const tag = (config.scopedEnvVar('NEW_VERSION_CHECK_TAG') as string | undefined) ?? 'latest'
 
   const checkVersion = async () => {
     try {
       // do not show warning if updating
-      if (process.argv[2] === 'update') return
-      const distTags = JSON.parse(await readFile(file, 'utf8'))
+      if (process.argv[2] === 'update' || this.config.scopedEnvVarTrue('SKIP_NEW_VERSION_CHECK')) return
+
+      const {contents: distTags, stat} = await readJSON<{[tag: string]: string}>(file)
+
+      // If the file was modified before the timeout, don't show the warning
+      if (
+        warningFrequency &&
```

Comment:
confirming that if `warningFrequency` is set to 0 you'd short-circuit the rest of the conditional?

Reply:
Yes, I think that makes sense. Frequency of `0` causes the default behavior of seeing the warning on every command execution.

I guess the alternative is that a frequency of `0` turns the warning off and you never see it again? We already have an env var for that so I don't see a reason to duplicate it